packages/mana/manabasevector.go

```diff
@@ -175,6 +164,75 @@ func (m *ManaBaseVector) Book(txInfo *TxInfo) {
 	}
 }
 
+func (m *ManaBaseVector) BookEpoch(created []*ledger.OutputWithMetadata, spent []*ledger.OutputWithMetadata) {
+	var revokeEvents []*RevokedEvent
+	var pledgeEvents []*PledgedEvent
+	var updateEvents []*UpdatedEvent
+	// only lock mana vector while we are working with it
+	func() {
+		m.Lock()
+		defer m.Unlock()
+		// first, revoke mana from previous owners
+		for _, output := range spent {
+			idToRevoke := m.getIDBasedOnManaType(output)
+			outputIOTAs, _ := output.Output().(devnetvm.Output).Balances().Get(devnetvm.ColorIOTA)
+			oldMana := m.getOldManaAndRevoke(idToRevoke, float64(outputIOTAs))
+			// save events for later triggering
+			revokeEvents = append(revokeEvents, &RevokedEvent{idToRevoke, float64(outputIOTAs), output.OutputMetadata().CreationTime(), m.Type(), output.ID().TransactionID, output.ID()})
+			updateEvents = append(updateEvents, &UpdatedEvent{idToRevoke, &oldMana, m.M.Vector[idToRevoke], m.Type()})
+
+		}
+		// second, pledge mana to new nodes
+		for _, output := range created {
+			idToPledge := m.getIDBasedOnManaType(output)
+			outputIOTAs, _ := output.Output().(devnetvm.Output).Balances().Get(devnetvm.ColorIOTA)
+			oldMana := m.getOldManaAndPledge(idToPledge, float64(outputIOTAs))
+			pledgeEvents = append(pledgeEvents, &PledgedEvent{idToPledge, float64(outputIOTAs), output.OutputMetadata().CreationTime(), m.Type(), output.Output().ID().TransactionID})
+			updateEvents = append(updateEvents, &UpdatedEvent{idToPledge, &oldMana, m.M.Vector[idToPledge], m.Type()})
+		}
+	}()
+
+	m.triggerManaEvents(revokeEvents, pledgeEvents, updateEvents)
+}
+
+func (m *ManaBaseVector) getIDBasedOnManaType(output *ledger.OutputWithMetadata) identity.ID {
+	var id identity.ID
+	if m.Type() == ConsensusMana {
+		id = output.OutputMetadata().ConsensusManaPledgeID()
+	} else {
+		id = output.OutputMetadata().AccessManaPledgeID()
+	}
+	return id
+}
+
+func (m *ManaBaseVector) getOldManaAndRevoke(oldPledgeNodeID identity.ID, amount float64) ManaBase {
+	if _, exist := m.M.Vector[oldPledgeNodeID]; !exist {
+		// first time we see this node
+		m.M.Vector[oldPledgeNodeID] = &ManaBase{}
+		m.M.Vector[oldPledgeNodeID].Init()
```

Comment:
Use `NewManaBase`, it is more idiomatic. Also it won't be necessary to call `Init`